Mathlib/Tactic/SuccessIfFailWithMsg.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Simon Hudon, Sébastien Gouëzel, Scott Morrison, Thomas Murrills
+-/
+import Lean
+
+/-!
+# Success If Fail With Message
+
+This file implements a tactic that succeeds only if its argument fails with a specified message.
+
+It's mostly useful in tests, where we want to make sure that tactics fail in certain ways under
+circumstances.
+-/
+
+open Lean Elab Meta Tactic Syntax
+
+namespace Mathlib.Tactic
+
+/-- `success_if_fail_with_msg msg tacs` runs `tacs` and succeeds only if they fail with the message
+`msg`.
+
+`msg` can be any term that evaluates to an explicit `String`. -/
+syntax (name := successIfFailWithMsg) "success_if_fail_with_msg " term:max tacticSeq : tactic
+
+/- Note that we don't use the following in our definition of `success_if_fail_with_msg`, because we
+want to check for definitional equality to the Expr specified by the syntax `$msg:term` there
+instead. We also can provide more descriptive error messages in the elaboration case. -/
```

Comment:
I'm confused, why are you checking the strings for definitional equality as expressions? Isn't string equality much faster?

Reply:
Hmm, what do you mean by this sentence?
> But even if you parse it as an expression, you should be able to evaluate it as a String.
Do you mean that we should use an unsafe evaluation? (Which wouldn't be the worst thing; I was just trying to avoid it if I could.)

In general, I think being able to define the strings elsewhere will be useful, because sometimes the error messages that we're checking for can be very long (and are sometimes reused in multiple cases), so it would help us see what a test is doing more easily!

The two options I can think of which would let us define error message strings are 1) parse as `str` if possible, and use `isDefEq` for everything that isn't a literal string 2) use an unsafe evaluation like `evalTerm`. Does one of those sound alright?

Reply:
Apologies for the delay! This should be fixed now. I also generalized it a bit and made the monadic version as general as possible (and refactored the tactic to use it).

Reply:
I think this comment still needs to be updated/removed?

Reply:
Oops, you're absolutely right. (I don't think any part of it is still relevant, nor does it particularly need a replacement, so I removed it.)